app/controllers/concerns/dynamic_field_names_cacher.rb

```diff
@@ -0,0 +1,21 @@
+module DynamicFieldTitlesCacher
+  CACHE_KEY = "%{identifier}/%{record_type}/column-keys/%{tail}".freeze
+
+  # Takes an ActiveRecord::Relation so we can make one more query off it
+  def collection_field_names(collection, identifier = "projects-#{@current_project.id}")
+    last_updated_record = collection.order(updated_at: :desc).first
+    # Exit early if the collection is empty.
+    return [] unless last_updated_record
```

Comment:
I'd make this L1 of the method, and replace with `collection.empty?` to make it clear it has nothing to do with the :last_updated value. Presumably the `collection` has already been loaded, so it won't add an extra SQL to just check for empty? (or any?)